lexers/mediawiki.lua

```diff
@@ -1,42 +1,72 @@
-local tag_start = token(lexer.TAG, '<' * P('/')^-1 * lexer.alnum^1 * lexer.space^0)
-local dq_str = '"' * ((lexer.any - S('>"\\')) + ('\\' * lexer.any))^0 * '"'
-local tag_attr = token(lexer.ATTRIBUTE, lexer.alpha^1 * lexer.space^0 *
-	('=' * lexer.space^0 * (dq_str + (lexer.any - lexer.space - '>')^0)^-1)^0 * lexer.space^0)
-local tag_end = token(lexer.TAG, P('/')^-1 * '>')
-lex:add_rule('tag', tag_start * tag_attr^0 * tag_end)
-
--- Link
-lex:add_rule('link', token(lexer.STRING, S('[]')))
-lex:add_rule('internal_link', B('[[') * token(lexer.LINK, (lexer.any - '|' - ']]')^1))
-
--- Templates and parser functions.
-lex:add_rule('template', token(lexer.OPERATOR, S('{}')))
+-- Internal Link: [[Target]] or [[Target|Display Text]]
+-- The content can contain almost anything except unbalanced square brackets.
+-- We'll highlight the whole thing as LINK.
+local internal_link_content = (lexer.any - P(']]'))^1 -- Matches everything until ']]'
+lex:add_rule('internal_link', lex:tag(lexer.LINK, P('[[') * internal_link_content * P(']]')))
+
+-- External Link: [http://example.com Link text] or [http://example.com]
+-- Content should start with a protocol (http/s, ftp, mailto etc.)
+local protocol = lexer.alpha^2 * P('://')
+local external_link_content = (protocol * (lexer.any - P(']'))^1) + (lexer.any - P(']'))^1
+lex:add_rule('external_link', lex:tag(lexer.LINK, P('[') * external_link_content * P(']')))
+
+-- Parser Functions: {{#function:args}} or {{function:args}}
+-- This is a very complex area. This lexer assumes a simple "name:" pattern.
+-- Tag the function name and its arguments.
+local parser_function_name = P('#')^-1 * (lexer.alpha + S('_'))^1 * P(':')
+local parser_function_content = (lexer.any - S('{}'))^1
 lex:add_rule('parser_func',
-	B('{{') * token(lexer.FUNCTION, '#' * lexer.alpha^1 + lexer.upper^1 * ':'))
-lex:add_rule('template_name', B('{{') * token(lexer.LINK, (lexer.any - S('{}|'))^1))
+	lex:tag(lexer.FUNCTION, P('{{') * parser_function_name * parser_function_content * P('}}')))
+
+-- Templates and Variables: {{TemplateName|args}} or {{VARIABLENAME}}
+-- Tag the template/variable name.
+-- This rule needs to be placed *after* parser_func if there's any ambiguity in parsing.
+local template_or_variable_name = (lexer.alnum + S('_'))^1
+local template_content = (P('|') * (lexer.any - S('{}'))^1)^-1 -- Optional content after |
+lex:add_rule('template',
+	lex:tag(lexer.VARIABLE, P('{{') * template_or_variable_name * template_content * P('}}')))
+
+-- Headings (e.g., == My Heading ==)
+-- Capture the heading text as lexer.HEADING
+local heading_level = S('=')^1
+lex:add_rule('heading',
+	lex:tag(lexer.HEADING, lexer.starts_line(heading_level * lexer.space^0 *
+		(lexer.any - S('=') - lexer.newline)^1 * lexer.space^0 * heading_level)))
 
 -- Operators.
-lex:add_rule('operator', token(lexer.OPERATOR, S('-=|#~!')))
+-- Consider adding more specific rules for bold/italic instead of general operators.
+-- For now, keep existing general operators.
```

Comment:
Since we're doing the migration, we may as well tag with `lexer.BOLD`, `lexer.ITALIC`, etc.